Introduce versioning for SDKs (#235)
meilisearch/integration-guides: resources/versioning.md

# Versioning policy

> This page describes the versioning rules Meilisearch SDKs will follow once Meilisearch engine v1.0.0 is released and how/when we should increase the MAJOR, MINOR, and PATCH of the versions.

Author: Good catch!

# Versioning policy

This page describes the versioning rules Meilisearch SDKs will follow once Meilisearch engine v1.0.0 is released and how/when we should increase the MAJOR, MINOR, and PATCH of the versions.

## 🤖 Basic rules

Meilisearch SDKs releases follow the [SemVer rules](https://semver.org/), including the following basic ones:

> 🔥 Given a version number MAJOR.MINOR.PATCH, increment the:
>
> 1. MAJOR version when you make incompatible API changes
> 2. MINOR version when you add functionality in a backwards compatible manner
> 3. PATCH version when you make backwards compatible bug fixes

**Changes that MAY lead the Meilisearch users (developers) to change their code are considered API incompatibility and will make us increase the MAJOR version of Meilisearch SDK.**

> **In other terms, if the users MAY have to do more steps than just downloading the new Meilisearch instance and running it, a new MAJOR is needed.**

Author: I've changed the wording :)

Examples of changes making the code break and then, involving increasing the MAJOR:

- Any change related to the public API the SDK exposes like:
    - Renaming a method name or classes
    - Removing or changing a positional argument of a method.

## Reasons why we should release a major version

#### New runtimes adoption

We don't have a strict policy yet regarding the bare minimum support of language version. Example:

Ruby SDK has a `required_ruby_version >= 2.6` which prevents users to use the gem if they are running their apps with a `Ruby 2.5`.

Is often good to only allow newer versions of the runtimes because they are maintained :sparkles: and there is always new features we can adopt in the code.

But, there is a price on that, when we make that cut, we reduce the amount of users that will be able to adopt our SDK/service.

List of current supported language version runtimes*:

| Language/Framework | Version |
| ------------------ | ------- |
| .NET | netstandard2.0 |
| Dart | >= 2.12.0 |
| Go | >= 1.16 |
| Java | >= 8.0 |
| JavaScript | N/A |
| Swift | >= 5.2 |
| PHP | >= 7.4 |
| Python | >= 3.7 |
| Ruby | >= 2.6 |
| Rust | N/A |

> When a SDK drops a minium version, we will release a new MAJOR version.

Author: On the JS side, for example, it is very common to mention this change as a breaking:

- https://github.com/ember-cli/ember-cli-htmlbars/releases/tag/v6.0.0
- https://github.com/mochajs/mocha/releases/tag/v10.0.0
- https://github.com/typed-ember/ember-cli-typescript/releases/tag/v5.0.0
- https://github.com/emberjs/ember-qunit/releases/tag/v6.0.0

> When a SDK drops a minium version, we will release a new MAJOR version.

Reviewer: It's as you wish then! 👌

\* This list may be outdated, always check the real version in the integration's repository.

> ## Correlation between Meilisearch engine versioning and the SDKs

Author: That's good you approached this topic because this is one of my doubts.

I would like to have this "justification and communication" written somewhere, and also rules to be followed. Where do you think they should go (the justification and communication)?

> ## Correlation between Meilisearch engine versioning and the SDKs

Reviewer: For me it's more about the form. You start by saying, "there is no plan to make SDK into v1"

This is not what the users want to know when they read your guide. They want to know 
- if the SDK versions will follow the version of Meilisearch or not 
- and then, which version of SDK is compatible with which version of Meilisearch, and how they can get the information

I would start to give your rules first, and then, the justifications, following by "we are open to feedback and your needs...blabla". Not the contrary

Currently there is no plan to make every SDK be in the v1 state for some reasons:

- Not enough feedback data to validate if the public API we expose to the users is the best.
- Ability to evolve every SDK individually without being attached to the engine version.

It means there is no guarantee that since Meilisearch engine is v1 we will have: Meilisearch Dart v1, Meilisearch JavaScript v1, Meilisearch .NET v1 and so on... 

> And also if we got a Meilisearch v2 in the future there is no guarantee that we will have Meilisearch Dart v2 as well.

Reviewer 2: ```suggestion
And also if we get a Meilisearch v2 in the future, there is no guarantee that we will have Meilisearch Dart v2 as well.
```

> And also if we got a Meilisearch v2 in the future there is no guarantee that we will have Meilisearch Dart v2 as well.

Author: After Charlotte's suggestions, I changed to:

> It means there is no guarantee that since Meilisearch engine is v1 we will have: Meilisearch Dart v1, Meilisearch JavaScript v1, Meilisearch .NET v1 and so on...
And also if we get a Meilisearch v2 in the future, there is no guarantee that we will have Meilisearch Dart v2 as well.

But yeah, it is just to inform the users, that they should not expect that.

What you can expect, is a table like this in every integration README:

| Meilisearch engine version | Integration version |
| ----------- | ----------- |
| v0.30 | v0.14.0 |

> | v1.0 | v0.15.0 or later |

Reviewer 2: `or` or `and` ?
```suggestion
| v1.0 | >= v0.15.0 |
```

| v1.1 | v0.18.0 |

> It means: The integration v0.15 will work with Meilisearch v1.1 running, but not all the new features introduced in the v1.1 will work in this integration version.

Reviewer 2: ```suggestion
Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but not all the new features introduced in the v1.1 will work with this integration version.
```

> ## When SDKs will be v1?

Reviewer 2: ```suggestion
## When to we update SDKs to v1?
```

> ## When SDKs will be v1?

Author: I added some content about it, maybe it now better explained.

There is no real answer for this question, but the accurated guess is, when we feel confortable with the current state of the integration.

> Or external facts that we may not control. So, if you want to only use v1 version of the SDKs please, open an issue in the integration-guides.

Reviewer 2: > Or external facts that we may not control

For example?

Author: Updated!

> Or external facts that we may not control. So, if you want to only use v1 version of the SDKs please, open an issue in the integration-guides.

Reviewer 2: ```suggestion
Or external facts that we may not control. If you only want to use a v1 version of a SDKs please, open an issue in the integration-guides. It helps us decide when to update to v1. 
```

## Important facts

> :warning: Most of the SDKs have methods like get `GetAllRawIndexesAsync`, `allRaw`, `rawSearch` and similar.
> 
> These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.

Reviewer 2: ```suggestion
:warning: Most of the SDKs have methods like get `GetAllRawIndexesAsync`, `allRaw`, `rawSearch` and similar. These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.
```

> have NO guarantee at all

Guarantee of what?

Author: I'll rework in the entire phrase, thanks for the highlight!

> These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.

Author: This is more a communication to the users about those methods, I would like to mention they are not "protected" by the layer of the SDK. My new version after Charlotte's comments is: WDYT?

----
:warning: Most of the SDKs have methods like get `GetAllRawIndexesAsync`, `allRaw`, `rawSearch` and similar.

These methods made the request directly to the Meilisearch engine and they also respond with the raw answer.
It means you should take care of possible inconsistencies that may appear in the Meilisearch engine response.

So, if possible don't use them.

> These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.

Reviewer: Hum I understand better, it's to do a warning saying "since these methods are only based on the Meilisearch response, ensure you use the expected version of Meilisearch".
But for me this is not a warning only for these kinds of methods. The warning is important but should include more cases. Indeed, a user trying to use a new feature, let's say "multi-index-search", by calling the `muli_index_search()` method but with the wrong version of Meilsearch, will have the same issue. The user should take care in general about using the right version of Meilisearch with the right SDK version.

> So, if possible don't use them.

I don't understand why, since Meilisearch will not break

> These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.

Reviewer 2: I think that the difference with the other methods is that we cannot introduce backward compatibility if we'd like to.

> These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.

Reviewer: Ok, then the guide needs to be more precise, because it's not obvious 😇 
however since we don't even have planned v2, I'm not sure this is an issue we should care of now.

> These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.

Author: Fixed!

> So, if possible don't use them.

Reviewer 2: ```suggestion
We suggest avoiding using them.
```

This page describes the versioning rules Meilisearch SDKs follow after the release of Meilisearch engine v1.0.0. It explains how and when the Integrations Team should increase the MAJOR, MINOR, and PATCH of the versions.

## 🤖 Basic rules

Meilisearch SDKs releases follow the [SemVer rules](https://semver.org/), including the following basic ones:

> 🔥 Given a version number MAJOR.MINOR.PATCH, increment the:
>
> 1. MAJOR version when you make incompatible API changes
> 2. MINOR version when you add functionality in a backwards compatible manner
> 3. PATCH version when you make backwards compatible bug fixes

**Changes that MAY lead the Meilisearch users (developers) to change their code are considered API incompatibility and will make us increase the MAJOR version of Meilisearch SDK.**

**In other terms, if the users MAY have to do more steps than just downloading the new Meilisearch package and running it, a new MAJOR is needed.**

Examples of changes making the code break and then, involving increasing the MAJOR:

- Any change related to the public API the SDK exposes like:
    - Renaming a method name or classes
    - Removing or changing a positional argument of a method.

> ## Reasons why releasing a major version is needed

Reviewer: I'm not sure I really understand the title
Plus there is one `##` title, and only one `###` title inside. Not sure the `##` title is needed.

Author: You're right, I'll remove the H2 title :)

> ### New runtimes adoption

Reviewer: Why `new`?

Author: What I wanted to mean with this were to imply that when we want to adopt a "new" runtime (different than the current supported ones), we should release a major (if we do the cut as explained). Maybe a better choice could be: `Minimum supported runtime`

> There is no strict policy yet regarding the minimum support of the language version. Example:

Reviewer: ```suggestion
There is no strict policy yet regarding the minimum support of the language version, but we have still defined some rules. Example:
```

Author: Since we didn't use the first person tense, I'll change to: "There is no strict policy yet regarding the minimum support of the language version, but some ground rules were defined. Example:"

Reviewer: Better indeed!

Ruby SDK has a `required_ruby_version >= 2.6`, which prevents users from using the gem if they run their apps with a `Ruby 2.5`.

It is often good to only allow newer versions of the runtimes because they are maintained :sparkles: and there are always new features is possible to adopt in the code and improve maintenance time.

> But there is a price for that. When that cut is made, the number of users who can adopt the SDK is reduced.

Reviewer: You can turn it differently to make it more user-oriented, but still clear for the integration team, something like "we always try to avoid dropping support because we know we are aware some users could be left behind"

Author: Good idea, done! I just changed the message itself to make it in third person!

| PHP | >= 7.4 |
| Python | >= 3.7 |
| Ruby | >= 2.6 |
| Rust | N/A |

When an SDK drops a minimum required version of the runtime, a new MAJOR version of the SDK should be released.

\* This list may be outdated. Always check the current version in the integration's repository.

## Correlation between Meilisearch engine versioning and the SDKs

SDKs do not follow the Meilisearch engine version. It means that there is no correlation between SDK's version and Meilisearch; an SDK may be in version `v8.0.0` while the Meilisearch engine is in `v1.0.0`.

Suppose the version of an SDK is the same as the latest version of Meilisearch. It is just a happy coincidence.

You can find out which version of the Meilisearch engine an SDK is compatible with by going to the `Compatibility with Meilisearch` section of every SDK README page. For example, the [README](https://github.com/meilisearch/meilisearch-ruby/#-compatibility-with-meilisearch) page of the ruby SDK.

To illustrate, every repository will have a similar table like this:

| Meilisearch engine version | Integration version |
| -------------------------- | ------------------- |
| v0.30 | v0.14.0 |
| v1.0 | >= v0.15.0 |

> | v1.1 | >= v0.18.0 |

Reviewer 2: It is indeed a bit confusing for the user. It does look like the user requires v0.18.0. 
Additionally, does it mean that, for example, javascript does not have a valid version for v1.0 as it is missing some features like the CSV document addition?

Maybe we should remove this table and avoid unnecessary complexity

- Any change related to the public API the SDK exposes like:
    - Renaming a method name or classes
    - Removing or changing a positional argument of a method.

## Reasons why we should release a major version

### New runtimes adoption

We don't have a strict policy yet regarding the minimum support of the language version. Example:

Ruby SDK has a `required_ruby_version >= 2.6`, which prevents users from using the gem if they run their apps with a `Ruby 2.5`.

It is often good to only allow newer versions of the runtimes because they are maintained :sparkles: and there are always new features we can adopt in the code.

But there is a price for that. When we make that cut, we reduce the number of users who can adopt our SDK/service.

Currently, all supported versions are set in the CI of each repository.

This is the list of minimum supported language version runtimes*:

| Language/Framework | Version |
| ------------------ | ------- |
| .NET | netstandard2.0 |
| Dart | >= 2.12.0 |
| Go | >= 1.16 |
| Java | >= 8.0 |
| JavaScript/Node | >= 14 |
| Swift | >= 5.2 |
| PHP | >= 7.4 |
| Python | >= 3.7 |
| Ruby | >= 2.6 |

> | Rust | N/A |

Reviewer 3: Does that mean we will have to update it each time we drop the support of a version?

Author: Yeah! But this kind of change does not happen in a high-frequency, may once a year or more?

When an SDK drops a minimum required version of the runtime, we release a new MAJOR version of the SDK.

\* This list may need to be updated. Always check the current version in the integration's repository.

## Correlation between Meilisearch engine versioning and the SDKs

SDKs do not follow the Meilisearch engine version. It means that there is no correlation between SDK's version and Meilisearch; an SDK may be in version `v10.0.0` while the Meilisearch engine is in `v1.0.0`.

Suppose the version of an SDK is the same as the latest version of Meilisearch. It is just a happy coincidence.

> A major release in Meilisearch may also not lead to a major release in the SDKs. For example, if we get a Meilisearch v2 in the future, there is no guarantee that the Meilisearch Dart SDK will update from v10 to v11.

Author: Cool, good to know, I was scared that what I said earlier was not comprehensible :)

You can find out which version of the Meilisearch engine an SDK is compatible with by going to the `Compatibility with Meilisearch` section of every SDK README page. For example, the [README](https://github.com/meilisearch/meilisearch-ruby/#-compatibility-with-meilisearch) page of the ruby SDK.

To illustrate, the idea is to have a similar table in every repository.

| Meilisearch engine version | Integration version |
| -------------------------- | ------------------- |
| v0.30 | v0.14.0 |
| v1.0 | >= v0.15.0 |
| v1.1 | v0.18.0 |

> Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but may only some of the new features introduced in v1.1 will work with this integration version.

Author: I changed from:

> Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but may only some of the new features introduced in v1.1 will work with this integration version.

to: 

> Given the table above, a user using Meilisearch v1.1 does not need to have the v0.18 version of its SDK.
Of course, some features will not be present, but they should still work.

WDYT?

> Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but may only some of the new features introduced in v1.1 will work with this integration version.

Reviewer: Are you sure you need to talk about this? Starting from Meilisearch v1.0, every SDK will be compatible with any version of Meilisearch >= v1.0
I'm not sure the users would understand that "it is compatible, but some features would be missing". From my point of view, if some features are missing, it's not really compatible

> Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but may only some of the new features introduced in v1.1 will work with this integration version.

Reviewer: It's part of the confusion I put [here](https://github.com/meilisearch/integration-guides/issues/237#issuecomment-1422986421)
You need to define what is "compatible" for you

> Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but may only some of the new features introduced in v1.1 will work with this integration version.

Author: I left a comment [here](https://github.com/meilisearch/integration-guides/issues/237#issuecomment-1431882992) where I expect we will agree with the term.
 
So, since the versioning table is not helpful, I'll remove it from this guide. And in the future, we can work on a better version where we can generate it using the code samples.

We're still working on making all of our SDKs stable, but first, we would like to evolve some questions, like having more feedback about the public API we expose to the users. We're eager to receive feedback on changes you'd like to see in some SDKs before their v1 release.

## When to we update SDKs to v1?

We are getting closer to a possible v1 in the SDKs, but there has yet to be a real roadmap.
If you have a specific use case where you need the SDKs in v1, please let us know through the issues.

## Important facts

:warning: Always check the Meilisearch engine version you are running and the version supported by your current SDK version.

> :warning: If you use a `raw` method like `GetAllRawIndexesAsync`, `allRaw`, `rawSearch`, and similar. Please double-check the Meilisearch version you're running.

Reviewer 3: Is there still some `all` like `getAll` somewhere in some SDK?

Author: I would have to check that, but they just have a bad naming they are not so harmful like the `raw` ones.
